Approval of the NonAdminBackupStorageLocation

mpryc opened the pull request:
Implements approvals for the NaBSL objects.

Feature which allow to approve NaBSL by the cluster admin prior to the BSL creation.

Depends-on: https://github.com/openshift/oadp-operator/pull/1633

## Why the changes were made

To implement approvals for the cluster admin as described in the https://github.com/migtools/oadp-non-admin/issues/204

Fixes https://github.com/migtools/oadp-non-admin/issues/220

## How to test the changes made

### Run simulation tests

### Disable the feature
1. Deploy oadp operator with the following PR: https://github.com/openshift/oadp-operator/pull/1633
2. Set the DPA to:
```yaml
  nonAdmin:
    enable: true
    requireAdminApprovalForBSL: false
```
3. Nothing should be changed from the user perspective, on the cluster in the openshift-adp namespaces there will be `NonAdminBackupStorageLocationRequest` objects with approved flag
4. Modifying the NaBSLRequest objects and seting them to anything else then approve will make those reverted back to approve state.

### Enable the feature
1. Deploy oadp operator with the following PR: https://github.com/openshift/oadp-operator/pull/1633
2. Set the DPA to:
```yaml
  nonAdmin:
    enable: true
    requireAdminApprovalForBSL: true
```
3. Creating NaBSL should result in the condition not satisfied and requires admin to approve the NaBSLRequest object
4. Update the corresponding NaBSLRequest and set the:
```yaml
spec:
  approvalDecision: approve
```
5. NaBSL should be created
6. Patch the NaBSLRequest and set the flag to:
```yaml
spec:
  approvalDecision: reject # or pending
```
7. The NaBSL should be updated to reflect the decision, corresponding Velero BSL and Secret in the openshift-adp namespace shoud be removed.

### Update the NaBSL spec
This change disallows to update NaBSL spec as a result test can be performed to update spec that was already created. Changes to the spec will result in the BSL to being removed from the cluster and the user informed.

mpryc commented on api/v1alpha1/nonadminbackupstoragelocationrequest_types.go:
```diff
@@ -0,0 +1,105 @@
+	ApprovalDecision NonAdminBSLRequest `json:"approvalDecision,omitempty"`
+}
+
+// VeleroBackupStorageLocationRequest contains information of the related Velero backup object.
```
Fixed & updated corresponding oadp-operator part in that project.

mateusoliveira43 commented on internal/predicate/nonadminbackupstoragelocationrequest_predicate.go:
```diff
@@ -0,0 +1,59 @@
+type NonAdminBackupStorageLocationRequestPredicate struct{}
+
+// Create event filter accepts all NonAdminBackupStorageLocationRequest create events
+func (NonAdminBackupStorageLocationRequestPredicate) Create(ctx context.Context, evt event.CreateEvent) bool {
```
can be deleted? only Update and Delete are used, right?

mpryc replied:
The delete causes the controller to re-create, so it is handled. This is to ensure we always have corresponding Request, including situations where controller is started later.

If admin Deletes the `NonAdminBackupStorageLocationRequest` then it has lost information about previous approval and it reverts back to `pending`, deleting corresponding BSL objects.

mateusoliveira43 commented on internal/handler/nonadminbackupstoragelocationrequest_handler.go:
```diff
@@ -0,0 +1,73 @@
+	// Create event handler for the NonAdminBackupStorageLocationRequest object
+}
+
+// Update event handler adds Velero Backup's NonAdminBackup to controller queue
```
typo

mpryc replied:
fixed.

mpryc commented on internal/predicate/nonadminbackupstoragelocationrequest_predicate.go:
```diff
@@ -0,0 +1,59 @@
+	return false
+}
+
+// Delete event filter accepts all NonAdminBackupStorageLocation delete events
```
fixed.

mpryc commented on internal/predicate/nonadminbackupstoragelocationrequest_predicate.go:
```diff
@@ -0,0 +1,59 @@
+	return false
+}
+
+// Update event filter only accepts NonAdminBackupStorageLocation update events that include spec change
```
fixed.

mateusoliveira43 commented on internal/common/function/function.go:
```diff
@@ -661,3 +682,34 @@ func ComputePrefixForObjectStorage(namespace, customPrefix string) string {
+// HasBSLChangesForApproval checks if there are any changes between the old and new BackupStorageLocationSpec
+// that would require admin approval. We do not want the admin to re-approve the BSL if it's just a matter of
+// the BSL creds being updated. In case of nil specs, we return true to trigger approval.
+func HasBSLChangesForApproval(oldSpec, newSpec *velerov1.BackupStorageLocationSpec) bool {
```
is this used?

mpryc replied:
Nice catch. It was firstly implemented when we had the update of NaBSL operation included in the design, however after talking to @shubham-pampattiwar and @weshayutin the idea changed to make things simpler and not allow update of the NaBSL and it's re-approval workflow.

Instead as in the updated design doc, the re-approval requires user to firstly delete NaBSL and then re-create new NaBSL with updated spec. It should be not a very often task to update NaBSLs by the users.

I've removed that function and relevant tests.

shubham-pampattiwar commented on internal/controller/nonadminbackupstoragelocation_controller.go:
```diff
@@ -377,6 +413,180 @@ func (r *NonAdminBackupStorageLocationReconciler) setFinalizerOnNaBSL(ctx contex
+		statusCondition = metav1.ConditionTrue
+		expectedPhase = nabsl.Status.Phase
+		terminalErr = reconcile.TerminalError(errors.New(message))
+	} else if nabslRequest.Status.VeleroBackupStorageLocationRequest.NACUUID == constant.EmptyString || nabslRequest.Status.VeleroBackupStorageLocationRequest.NACUUID != nabsl.Status.VeleroBackupStorageLocation.NACUUID {
```
++

shubham-pampattiwar commented on internal/controller/nonadminbackupstoragelocation_controller.go:
```diff
@@ -377,6 +413,180 @@ func (r *NonAdminBackupStorageLocationReconciler) setFinalizerOnNaBSL(ctx contex
+	} else {
+		switch {
+		case nabslRequest == nil:
+			shouldRequeue = true
```
I think you can remove this case, right ?
This condition will always be false (GetNabslRequestByLabel will return an err in nil case)

mpryc replied:
Correct, actually `GetNabslRequestByLabel` is not returning error if nothing is found, works same in other Backup/Restore/NAB objects so don't want to change how the GetNabslRequestByLabel works, however this check is not required and added new one on the nabslRequest nil just afte GetNabslRequestByLabel so other parts of the function won't have nil object (which will in theory never happen, but better safe then sorry)

shubham-pampattiwar commented on internal/controller/nonadminbackupstoragelocation_controller.go:
```diff
@@ -377,6 +413,180 @@ func (r *NonAdminBackupStorageLocationReconciler) setFinalizerOnNaBSL(ctx contex
+
+	updatePhase := updateNonAdminPhase(&nabsl.Status.Phase, expectedPhase)
+
+	if statusCondition == metav1.ConditionFalse {
```
Should we also check the type of condition here ?

mpryc replied:
No need as we have full control over this variable and we set this variable on line:
https://github.com/migtools/oadp-non-admin/pull/216/files/30d565184e6f17c9f1a64af684ed86f2eed90018#diff-ab1df5cab3611d202a4b22ab31a471f15bd9b010c803a8df03b55ce6b7e09c2eR433

So it's not value which we get from the cluster from the condition, but we know for which condition we are setting it to true/false.

I've renamed this var to be more informative: `adminApprovedCondition`

shubham-pampattiwar reviewed (commented):
Added a few comments, overall the flow looks good ! Thanks @mpryc !

mpryc commented on api/v1alpha1/nonadminbackupstoragelocationrequest_types.go:
```diff
@@ -0,0 +1,105 @@
+
+// Predefined NonAdminBSLRequestConditions
+const (
+	NonAdminBSLRequestApproved NonAdminBSLRequest = "approve"
```
I structured the action conditions as actionable commands to reflect the action:

 - (I) approve
 - (I) reject
 - pending <-- action to be taken

The corresponding phase names are in the past tense to reflect state (after reconcile approve spec is updated to phase status "Approved":

- Approved
- Rejected
- Pending

I can change it or we can keep the above rationale ?

shubham-pampattiwar replied:
I see, no problem, I agree with current changes.

mpryc commented on api/v1alpha1/nonadminbackupstoragelocationrequest_types.go:
```diff
@@ -0,0 +1,105 @@
+)
+
+// NonAdminBackupStorageLocationRequestSpec defines the desired state of NonAdminBackupStorageLocationRequest
+type NonAdminBackupStorageLocationRequestSpec struct {
```
It's in the Status, so admin can not simply generate the approval prior to it's existence - this is to reduce risk of improper labels. This spec isn't meant to be created from scratch.

mpryc commented on api/v1alpha1/nonadminbackupstoragelocationrequest_types.go:
```diff
@@ -0,0 +1,105 @@
+// NonAdminBackupStorageLocationRequestStatus defines the observed state of NonAdminBackupStorageLocationRequest
+type NonAdminBackupStorageLocationRequestStatus struct {
+	// +optional
+	VeleroBackupStorageLocationRequest *VeleroBackupStorageLocationRequest `json:"veleroBackupStorageLocationRequest,omitempty"`
```
About Name/Namespace - comment above, we have it here: https://github.com/migtools/oadp-non-admin/pull/216/files/30d565184e6f17c9f1a64af684ed86f2eed90018#diff-848168a85769ada3f4d261587d865e7391289a884a6a390cce8c6c6c0650740aR63-R70

This structure looks like:
 1. See nacuuid, name and namespace - it's there and it's inline with other statuses NAB or NAR - same name/namespace/nacuuid string names.
 2. The `veleroBackupStorageLocationRequest` could be just `nonAdminBslRequest` to make it shorter or remove it all together and put all of the objects  ?

```yaml
spec:
  approvalDecision: pending
status:
  phase: Pending
  veleroBackupStorageLocationRequest:
    nacuuid: non-admin-bsl-test-example-c7a90b05-9a1b-4e4c-a1e8-c3ab7f2fd4de
    name: example
    namespace: non-admin-bsl-test
    requestedSpec:
      config:
        region: eu-central-1
      credential:
        key: mycred
        name: cloud-credentials
      objectStorage:
        bucket: mpryc-nac-second
        prefix: default-nac-bucket
      provider: aws
```

shubham-pampattiwar replied:
yeah I would use `nonAdminBSLRequest` because you have data in struct which pertains to NABSL and not VBSL

mpryc replied:
Actually NaBSLRequest is very confusing once I made this change.

It's because the NaBSLRequest is the object itself (Request object) and here in the status we are refering to nonAdminBsl object spec which was origin for our NaBSLRequest.

Maybe we call it full name, so it will look like

```yaml
spec:
  approvalDecision: pending
status:
  phase: Pending
  nonAdminBackupStorageLocation:
    nacuuid: non-admin-bsl-test-example-c7a90b05-9a1b-4e4c-a1e8-c3ab7f2fd4de
    name: example
    namespace: non-admin-bsl-test
    requestedSpec:
      config:
        region: eu-central-1
      credential:
        key: mycred
        name: cloud-credentials
      objectStorage:
        bucket: mpryc-nac-second
        prefix: default-nac-bucket
      provider: aws
```

mpryc replied:
@shubham-pampattiwar with the latest modification where I chosen `nonAdminBackupStorageLocation` the NonAdminBackupStorageLocationRequest object is rendered as follows, let me know if we want to change it or keep it:

```yaml
[...]
spec:
  approvalDecision: pending
status:
  nonAdminBackupStorageLocation:
    nacuuid: non-admin-bsl-test-nonadmi-09d36eb9-381f-4cad-bef6-1a354af58535
    name: nonadminbackupstoragelocation-sample
    namespace: non-admin-bsl-test
    requestedSpec:
      config:
        region: eu-central-1
      credential:
        key: mycred
        name: cloud-credentials
      objectStorage:
        bucket: mpryc-nac-second
        prefix: default-nac-bucket
      provider: aws
  phase: Pending
```

shubham-pampattiwar replied:
`nonAdminBackupStorageLocation` sounds good !

mpryc commented on internal/controller/nonadminbackupstoragelocation_controller.go:
```diff
@@ -377,6 +413,180 @@ func (r *NonAdminBackupStorageLocationReconciler) setFinalizerOnNaBSL(ctx contex
+		}
+
+		if !r.RequireApprovalForBSL && nabslRequest.Spec.ApprovalDecision != nacv1alpha1.NonAdminBSLRequestApproved {
+			logger.V(1).Info("NonAdminBackupStorageLocationRequest already exists and is not approved, updating to approved")
```
yes

mpryc replied:
Updated

mpryc commented:
@shubham-pampattiwar I think I addressed all of your comments and updated oadp-operator, please check them. Some of the comments I've implemented slightly different then you asked and gave rationale behind my implementation changes.

mateusoliveira43 commented on internal/controller/nonadminbackupstoragelocation_controller.go:
```diff
@@ -377,6 +413,180 @@ func (r *NonAdminBackupStorageLocationReconciler) setFinalizerOnNaBSL(ctx contex
+			Reason:  "BslSpecUpdateRejected",
+			Message: message,
+		})
+		adminApprovedCondition = metav1.ConditionTrue
```
since this error case, shouldn't it be false and BackingOff?

mpryc replied:
No.

In this scenario we have two conditions:

```go
	NonAdminBSLConditionApproved           NonAdminBSLCondition = "ClusterAdminApproved"
	NonAdminBSLConditionSpecUpdateApproved NonAdminBSLCondition = "SpecUpdateApproved"
```

This flag however is not used for updating any of the above conditions from (line number 480), it is set to true to not delete previously created BSL nor Secret on line number 490. This was added here and I can introduce new flag e.g. `deleteNacBslSecret`, but thought to use it to not have too many vars in this logic

mpryc replied:
added new flag `preserveVeleroBslSecret` in the go code, to make it easier to read the workflow. Thanks.

mpryc replied:
https://github.com/migtools/oadp-non-admin/pull/216/commits/3cf8e01c6a3d93bc6c0376fc390d136b6d2132af

mateusoliveira43 commented:
/hold for OADP PR

mpryc commented:
>   would remove it or change message to something like `automatically approved`

We would add the extra logic to distinguish between different approval types (human & non-human) which would add complexity in some cases when one is switched to another. Plus it would expose DPA setting to the users with this message, so I think either we think of better generic wording or stick with what's there.

> * if admin changes from no approval to approval, nothing happens. Should all NABSL be set to pending approval? (The other way around, when admin changes from approval to no approval, seems to work, i.e., all NABSL go to approved state)

This is actually desired. If the admin want's to re-approve all of the NaBSLs after toggling this to true, then the easiest is to remove all of the NonAdminBackupStorageLocationRequest objects from the openshift-adp namespace:
[ - If there was no approval required, switching it to on will NOT delete all of the previously approved 
](https://github.com/migtools/oadp-non-admin/blob/3cf8e01c6a3d93bc6c0376fc390d136b6d2132af/docs/design/non_admin_backupstoragelocation.md#enabling-bsl-approval-request-feature)

Vice-versa:
https://github.com/migtools/oadp-non-admin/blob/3cf8e01c6a3d93bc6c0376fc390d136b6d2132af/docs/design/non_admin_backupstoragelocation.md#disabling-bsl-approval-request-feature

shubham-pampattiwar commented on api/v1alpha1/nonadminbackupstoragelocationrequest_types.go:
```diff
@@ -0,0 +1,105 @@
+// NonAdminBackupStorageLocationRequestStatus defines the observed state of NonAdminBackupStorageLocationRequest
+type NonAdminBackupStorageLocationRequestStatus struct {
+	// +optional
+	NonAdminBackupStorageLocationRequestStatusInfo *NonAdminBackupStorageLocationRequestStatusInfo `json:"nonAdminBackupStorageLocation,omitempty"`
```
I though we wanted this as `nonAdminBackupStorageLocation` ?

mpryc replied:
Including structure of `NonAdminBackupStorageLocation` will look wird imo, because it contains yet another status which contains Velero BSL:
https://github.com/migtools/oadp-non-admin/blob/master/api/v1alpha1/nonadminbackupstoragelocation_types.go#L40

mpryc replied:
That's just go var name, which isn't used in the api/v1alpha1, right? If I used the `NonAdminBackupStorageLocation` then we had this object already https://github.com/migtools/oadp-non-admin/blob/121987f57ecb24d4bc65e4a2bef81593525a5a64/api/v1alpha1/nonadminbackupstoragelocation_types.go#L74

shubham-pampattiwar commented on internal/controller/nonadminbackupstoragelocation_controller.go:
```diff
@@ -377,6 +413,182 @@ func (r *NonAdminBackupStorageLocationReconciler) setFinalizerOnNaBSL(ctx contex
+		}
+
+		if !r.RequireApprovalForBSL && nabslRequest.Spec.ApprovalDecision != nacv1alpha1.NonAdminBSLRequestApproved {
+			logger.V(1).Info("Unapproved NonAdminBackupStorageLocationRequest found; approving as requireApprovalForBSL on the DPA is not true.")
```
++

shubham-pampattiwar commented:
/lgtm

weshayutin commented:
VERY WELL DONE FOLKS.. this is working for me as expected.
I have not checked for dupes yet.. but created two issues

https://github.com/migtools/oadp-non-admin/issues/232
https://github.com/migtools/oadp-non-admin/issues/231
